### ISSUE TITLE
flip with `boundary` gets clipped when it can't fit within `rootBoundary`

### ISSUE THREAD

**Issue by smhigley** [floating-ui/floating-ui#3014]
**Describe the bug**

A floating element with `flip({ boundary: document.body, roootBoundary: 'viewport' })` is ending up positioned so it is clipped by the document (or any other custom clipping boundary) when it cannot fit within the viewport.

**To Reproduce**
[Codesandbox link](https://codesandbox.io/p/sandbox/floating-ui-dom-template-forked-g958sd)

The codesandbox shows two examples -- one where the clipping ancestor is `document.body`, and one within a custom scroll region.

For the one within a scroll region, it is positioned above the trigger and clipped by the scroll boundary when scrolled all the way to the top.

For the one where the clipping boundary is `document.body`, to repro do the following:
- Set the viewport height to 225px or less
- scroll all the way to the top
- the floating element is clipped by the document:
<img width="633" alt="Screenshot of a button with a block of white text against a black background above it. The text is cut off at the top." src="https://github.com/user-attachments/assets/6ec2088f-8775-4ccc-bbe2-dcc93fb7b711">

**Expected behavior** 

Ideally I'd like the floating element to position itself to the best fit based on the viewport, unless that position causes it to be clipped by the document. From the docs, I thought `boundary` would do that, but please let me know if I'm misunderstanding what that prop is intended to do.

**Context:**

- OS: macOS 14.6
- Browser Chrome, Edge, Firefox, Safari (same behavior on all 4)

**Comment by atomiks** [comment]
It sounds like you just want to use `rootBoundary: 'document'` and keep the default `boundary` setting. This bases it on the document (basically `document.body` but more robust) instead of the viewport.

**Comment by smhigley** [comment]
@atomiks I think the issue with that approach is that we still have a pretty hard requirement for the floating element to use a best-fit positioning approach based on the visual viewport.

I think maybe a better description of the issue is that we're going for a pretty bog-standard `flip()` behavior, but we're running into issues when both the following are true:
- the floating element is too big to fit within the viewport on any side of the anchor
- the anchor is close to the top of the document

If both of those are true, the floating element positions itself above the anchor, and is fully inaccessible instead of positioning below the anchor, which would allow it to be accessed with scroll.

I didn't add this to the original issue b/c it didn't seem entirely relevant, but for context this is coming up as an accessibility issue, since with high zoom levels quite a lot of floating elements cannot fit within the viewport.

**Comment by atomiks** [comment]
Yeah I understood that as well. Usually you'd use the `size()` middleware so it can stay scrollable (the floating element itself will be scrollable in any direction, rather than needing to rely on an outer container's scroll)

**Comment by smhigley** [comment]
@atomiks thanks for the suggestion! We actually do use that middleware in some cases to ensure the floating element stays within the viewport.

The reason we don't use it all the time is because for some cases, having the additional scrollbar inside the floating element in addition to the page scrollbar is either less accessible or entirely inaccessible. So while `size()` works well for things like menus and comboboxes using floating UI, it doesn't work for these cases:
- Floating popups that are on the wider side, such that they could be close to the width of the screen when the screen is zoomed in. Doing this makes it hard to separately control scrolling for both the popup and the page, and this is compounded when the user is using assistive tech like zoom software, voice control, a switch setup (the assistive tech, not the nintendo one 😅), etc.
- When the floating surface does not take focus and does not have any focusable elements. When that is the case, it is impossible for keyboard & screen reader users to scroll it.

Do you know if `boundary` is intended to define an override to `rootBoundary` & force elements to stay within it in a scenario like this? I wasn't entirely sure based on the docs, but I also couldn't think of something else it was supposed to do. That is, I wasn't sure if this was a bug with `rootBoundary` or if I was just mistaken about that prop's intended purpose 😅.

**Comment by atomiks** [comment]
> Do you know if boundary is intended to define an override to rootBoundary & force elements to stay within it in a scenario like this?

`boundary` and `rootBoundary` act together to calculate the actual boundary, which is the maximum viewable area that's contained within both. So if the bottom of the `boundary` is below the bottom of the `rootBoundary`, the bottom of the `rootBoundary` is used instead (since that's the actual "viewing area"), etc. Though it does have a bias to top/left of each, instead of the respecting the `boundary` when it should: https://github.com/floating-ui/floating-ui/issues/2784

> When the floating surface does not take focus and does not have any focusable elements. When that is the case, it is impossible for keyboard & screen reader users to scroll it.

I think the floating element or scrollable container should always take focus in this scenario, otherwise it suggests it has a really long string of uninterrupted content via `aria-labelledby` / `aria-describedby` instead of breaking it up into individual paragraphs to read at the user's pace with the virtual cursor.

But regardless, I think creating a bias toward the direction of scroll is a valid algorithmic improvement over the current approach, which favors the best-fitting placement within the viewable area without considering if one of the sides can be scrolled.

**Comment by smhigley** [comment]
@atomiks re: the always-focusable popup -- it's often true that popups should take focus, but there are some notable exceptions --
- At a small effective viewport (through zoom, a small device, or both), especially if paired with text size increases, you don't actually need all that much content to exceed the available viewport space, especially if the trigger is near the middle.
- Having an image in the popup
- Page previews, similar to what you get when hovering links on Wikipedia or Github. You generally do not want to have these take focus, even though they may contain nested interactive items. You'd need a mechanism to get focus from the trigger to the popup since enter presumably triggers the link, and trying to build in direct access to the popup is generally less accessible/usable than just following the link (depending on context and the purpose of the preview).

Do you think this is something the team might want to pick up? Either WRT using `boundary` when `rootBoundary` has no fitting position, or just not allowing popups to render outside the top/left of `document.body`? Those seem maybe worth special-casing, since rendering outside of the right/bottom bounds will give the user scrollbars, while top/left will just render it fully inaccessible.
